### PR TITLE
Enable PDF upload to Digital Ocean Spaces

### PR DESCRIPTION
**Pull request by danielSbastos**
Hi,
This is my attempt to enable upload files to Digital Ocean Spaces in production.

The files are upload using `boto3` API for AWS S3, given that it is compatible with Digital Ocean Spaces.

In order to enable this functionality, you need to set the following environment variables

```sh
FILES_STORE=''
SPACES_NAME=your-spaces-name
SPACES_ACCESS_KEY_ID=your-key
SPACES_SECRET_ACCESS_KEY=your-key
```

 - `FILES_STORE`: if set to false or an empty string, Scrapy does not save the data locally.
 - `SPACES_ACCESS_KEY_ID`: an access key obtained in the `API` section your Digital Ocean Dashboard
 - `SPACES_SECRET_ACCESS_KEY`: an secret access key obtained in the `API` section your Digital Ocean Dashboard

 An example would be the following:

 ```sh
SPACES_NAME=test-scrapy
SPACES_ACCESS_KEY_ID=T3JRHS6M4CFIUH3EM62K
SPACES_SECRET_ACCESS_KEY=p4arD3AxCupbZxK4Iz/nlHAABaarIEoC4YQdD9G0Tl2
 ```

### CODE REVIEW

**Comment by danielSbastos** [comment]
https://github.com/okfn-brasil/diario-oficial/issues/16

**Comment by chicocvenancio** [comment]
Should we place the needed variables in the .env.example file, instead of using the bash command to export it?

**Comment by danielSbastos** [comment]
@chicocvenancio, no, that can't be done, not the way it is structured now, at least. If left on the `.env.example` file, these variables that should only apply to production will override the configuration to download the files to local storage in development and instead upload them to GSC.

**Comment by cuducos** [comment]
> If left on the .env.example file, these variables that should only apply to production will override the configuration to download the files to local storage in development and instead upload them to GCS.

Hi @danielSbastos. Either I got you wrong, or I agree with @chicocvenancio… but I'm afraid I'm with Chico.

The `.env` (and consequentially the `.env.example`) are the right tools to manage settings that should be one thing in dev environment and another thing in the production environment.

I do think need to add  `FILES_STORE`, `GCS_PROJECT_ID` and `GOOGLE_APPLICATION_CREDENTIALS` to the `.env`/`.env.exampe` as a way to make it clear that these are config stuff, [separated from the logic of the app](https://12factor.net/config).

For the sake of clarification:

* `.env.example`, as the name read, is a mere example of settings needed in order to spin up the app (no sensible data here); this is committed and part of the repository as reference for users;
* `.env` has the real data, sensible stuff such as authentication, credential and API keys; and this is never committed; users copy `.env.example` and edit according to their needs.

So, yes… I do believe we could move  `FILES_STORE`, `GCS_PROJECT_ID` and `GOOGLE_APPLICATION_CREDENTIALS` to the `.env`/`.env.exampe` scheme ; )

**Comment by danielSbastos** [comment]
@cuducos @chicocvenancio, I think I got it. Add in `.env.example` the `FILES_STORE`, `GCS_PROJECT_ID` and `GOOGLE_APPLICATION_CREDENTIALS`  but with a fake value. Then, when users copy it they alter the values just like the rest of vars. If in development mode, an empty string or nothing could be passed as the value whereas in production we define these variables. 

(@chicocvenancio, I really didn't know much about docker up until a few days ago, so the difference `docker-compose run` and `docker-compose up` was unknown to me and the issue with the env vars even less, sorry about that.)

I'll try later on today or tomorrow and will get back to you guys. Thanks for help :)

**Comment by danielSbastos** [comment]
@cuducos and @chicocvenancio, I sent two new commits, you guys can check them out and review again :)

**Comment by JaTvoiRabotnik** [comment]
What's the status of this pull request?

**Comment by cuducos** [comment]
@JaTvoiRabotnik, the PR looks good in general.

I haven't had time for a proper code review though. There's also something pending: I need to check with @irio why we opted for Google Cloud (the project is now sponsored by DigitalOcean and I think it would make sense to use DigitalOcean Spaces instead; at least we would be able to use credits we already have). If that's the case we might need to ask for some changes, or merge this to another feature branch in order to adapt it to DigitalOcean Spaces.

**Comment by danielSbastos** [comment]
@cuducos, we can close this PR (if it's no longer needed) and I'll be happy to try to implement remote storage using Digital Ocean Spaces.

**Comment by cuducos** [comment]
@danielSbastos, just to be sure: I don't think close this PR is the best option. I think that working with DigitalOcean Spaces should not be so different from Google Cloud buckets, so probably just minor changes to this PR would be enough : )

**Comment by danielSbastos** [comment]
@cuducos, sure, I agree with you! The only thing that worries me is that Scrapy file's pipeline does not support DigitalOcean Spaces, so I'll have to research a bit and try to implement it or find another solution. I'll get this going this week

**Comment by cuducos** [comment]
>  Scrapy file's pipeline does not support DigitalOcean Spaces

Does it support AWS S3? It's exactly the same API (you can even use `boto3` to write/read files in a DigitalOcean Spaces).

**Comment by danielSbastos** [comment]
@cuducos, it does support S3. Didn't know DigitalOcean Spaces API could be used on AWS S3's one, awesome. I will give try to implement it then  :)

**Comment by cuducos** [comment]
>  it does support S3

Would you mind expanding on that? I mean… I do understand there should be some incompatibility, but [S3 is mentioned in the same docs as Google Cloud in Scarpy docs](https://doc.scrapy.org/en/latest/topics/media-pipeline.html#amazon-s3-storage) so I'm a bit confused about what exactly might be missing here…

**Comment by danielSbastos** [comment]
DigitalOcean Spaces is built on top of AWS S3, such that by just changing the endpoint url to one of DigitalOcean, we can upload files to it. The docs say the following:

> The Spaces API aims to be interoperable with Amazon's AWS S3 API. In most cases, when using a client library, setting the "endpoint" or "base" URL to ${REGION}.digitaloceanspaces.com and generating a Spaces key to replace your AWS IAM key will allow you to use Spaces in place of S3.

However, given your question and little more of research, now I think it won't be as easy as replacing the 
`IMAGES_STORE = 's3://bucket/images'` to match one of DigitalOcean Spaces', the reason being that we would need to change the `ENDPOINT_URL` also. 

Let me build on top of this a bit; AWS S3 has endpoints by regions, for example, `s3.us-east-2.amazonaws.com` and `s3.us-west-2.amazonaws.com`, like on the quote above, DigitalOcean Spaces endpoint differs. We need somehow inherit the media pipelime s3 class and set the `ENDPOINT_URL` to point to DigitalOcean Spaces';

```py
class S3FilesStore(object):
    AWS_ACCESS_KEY_ID = None
    AWS_SECRET_ACCESS_KEY = None
    AWS_ENDPOINT_URL = None
    AWS_REGION_NAME = None
    AWS_USE_SSL = None
    AWS_VERIFY = None
..
```

```py
s3_store = S3FilesStore(...)
s3_store.AWS_ENDPOINT_URL = "<REGION>.digitaloceanspaces.com"
```

With this, I think another class needs to be created to deal with the nuances between S3 and Spaces. If in `production`, then we instanciate it and don't use the `IMAGES_STORE`.

I am 100% sure this will work, because as the docs say, 
> **In most cases**, when using a client library...,
but we can try.

**Comment by cuducos** [comment]
> We need somehow inherit the media pipelime s3 class and set the `ENDPOINT_URL` to point to DigitalOcean Spaces

Sounds like a great strategy! I would inherit from the s3 class, though:

```python
from scrapy.pipelines.files import S3FilesStore

class SpacesFilesStore(S3FilesStore):
    def __init__(self, *args, **kwargs):
        region = kwargs.pop('digitalocean_region')
        super().__init__(*args, **kwrags)
        self.AWS_ENDPOINT_URL = f"{digitalocean_region}.digitaloceanspaces.com"
```

And then use this class instead of:

```python
s3_store = S3FilesStore(...)
s3_store.AWS_ENDPOINT_URL = "<REGION>.digitaloceanspaces.com"
```

Does it make sense?

**Comment by danielSbastos** [comment]
@cuducos, yup, makes perfect sense, even better!

**Comment by rennerocha** [comment]
As you are using the `FilesPipeline`, you can define the endpoint URL in the settings file:
https://github.com/scrapy/scrapy/blob/master/scrapy/pipelines/files.py#L314

I believe that just including `AWS_ENDPOINT_URL` in your settings.py file should do the trick and it will not be necessary to create a new class. Please let me know if it works :-)

**Comment by danielSbastos** [comment]
@rennerocha, oh, okay. I'll give it a try, thanks!

**Comment by cuducos** [comment]
@danielSbastos what about updating Scrapy in this repo? As it's not a major bump, probably we're not going to break anything ; )

**Comment by danielSbastos** [comment]
@cuducos, sure, that's a good option.

**Comment by rennerocha** [comment]
I would suggest to upgrade Scrapy version to 1.5.1 too.

**Comment by danielSbastos** [comment]
I updated Scrapy but still wasn't able to upload. Digital Ocean Stores files stores does not have an `s3://`, however, the S3Pipeline requires this and is raising an error in case not. I think it'll be faster and easier to just create another pipeline and do the tweaks there.

File: processing/data_collection/gazette/service/spaces.py [hunk 1]
```diff
@@ -16,4 +16,4 @@ def __init__(self):
         )
 
     def upload_file(self, file_path):
-        self.client.upload_file(file_path, 'test-scrapy', file_path)
+        self.client.upload_file(file_path, os.getenv('SPACES_NAME'), file_path)
```
[the remark this answers is not in the document]
Comment (danielSbastos) [reply processing/data_collection/gazette/service/spaces.py]: Yes, sure, I will do this

**Comment by danielSbastos** [comment]
I will send another commit fixing the failing build due to `black`.

**Comment by willianpaixao** [comment]
@danielSbastos @cuducos do you mind if I make the final changes?

**Comment by danielSbastos** [comment]
go ahead, @willianpaixao :)

**Comment by jvanz** [comment]
Closing this one. There is the `QueridoDiarioFilesPipeline` class which uses the scrapy `FilesPipeline` to upload the files to Digital Ocean spaces or any s3ish storage.

**Pull request closed without merging**